rust/wal3/src/shard_manager.rs

```diff
@@ -0,0 +1,221 @@
+
+#[derive(Debug)]
+pub struct ShardManager {
+    options: ThrottleOptions,
+    state: Mutex<ManagerState>,
+    records_written: AtomicUsize,
+    batches_written: AtomicUsize,
+    write_finished: tokio::sync::Notify,
+}
+
+impl ShardManager {
+    pub fn new(options: ThrottleOptions, initial_manifest: &Manifest, shards: usize) -> Self {
+        let shards = (1..=shards)
+            .map(|idx| {
+                let shard_id = ShardID(idx);
+                let next_seq_no = initial_manifest
+                    .next_seq_no_for_shard(shard_id)
+                    .unwrap_or(ShardSeqNo(1));
+                let next_write = Instant::now();
+                ShardState {
+                    shard_id,
+                    next_seq_no,
+                    next_write,
+                    writers_active: 0,
+                }
+            })
+            .collect();
+        Self {
+            options,
+            state: Mutex::new(ManagerState {
+                shards,
+                enqueued: Vec::new(),
+                last_batch: Instant::now(),
+            }),
+            records_written: AtomicUsize::new(100_000),
+            batches_written: AtomicUsize::new(1),
+            write_finished: tokio::sync::Notify::new(),
+        }
+    }
+
+    pub fn push_work(
+        &self,
+        stream_id: StreamID,
+        message: Message,
+        tx: tokio::sync::oneshot::Sender<Result<LogPosition, Error>>,
+    ) {
+        // SAFETY(rescrv): Mutex poisoning.
+        let mut state = self.state.lock().unwrap();
+        state.enqueued.push((stream_id, message, tx));
+    }
+
+    pub async fn wait_for_writable(&self) {
+        self.write_finished.notified().await;
+    }
+
+    pub fn until_next_time(&self) -> Duration {
+        // SAFETY(rescrv): Mutex poisoning.
+        let state = self.state.lock().unwrap();
+        let elapsed = state.last_batch.elapsed();
+        let threshold = Duration::from_micros(self.options.batch_interval as u64);
+        if elapsed > threshold {
+            Duration::ZERO
+        } else {
+            threshold - elapsed
+        }
+    }
+
+    #[allow(clippy::type_complexity)]
+    pub fn take_work(
+        &self,
+        manifest_manager: &ManifestManager,
+    ) -> Option<(
+        ShardID,
+        ShardSeqNo,
+        LogPosition,
+        DeltaSeqNo,
+        Vec<(
+            StreamID,
+            Message,
+            tokio::sync::oneshot::Sender<Result<LogPosition, Error>>,
+        )>,
+    )> {
+        // SAFETY(rescrv): Mutex poisoning.
+        let mut state = self.state.lock().unwrap();
+        if state.enqueued.is_empty() {
+            return None;
+        }
+        let batch_size = self.batch_size();
+        let mut batch_size =
+            if state.enqueued.len() > batch_size * 2 || state.enqueued.len() < batch_size {
+                state.enqueued.len()
+            } else {
+                batch_size
+            };
```

Comment:
May I ask why is that batch size equal to the size of `enqueued` when it is larger than two times `batch_size`?

Reply:
Left a comment explaining this.